README.MD
![logo](docs/img/sprites.js.png)


JavaScript lightweight and simple in usage library that helps animating canvas sprites. What makes sprite.js so special is ability to controll frame rate without using `setTimeout()` or `setInterval()`. Visit [Demo page](https://wisniewski94.github.io/sprites.js/).

**774 bytes** gzipped. **1,56KB** minified.

## Getting Started
### 1. Download
You can fork or download the plugin from GitHub, or you can install it through npm or bower.

```
$ npm install sprites.js 
```

```
$ bower install sprites.js
```

### 2. Insert required files
In the page's footer, just before `</body>`, include Javascript files:
```html
<script src="sprites.min.js"></script>
<script src="custom.js"></script>
```
### 3. Create the HTML markup
Inside page's body place canvas with custom id. Ex.:

```
<div>
    <canvas id="my-sprite"></canvas>
</div>
```

## Usage
Inside your custom .js file create new object which will represent your sprite:

```javascript
var example = new Sprite({
    src: 'img/papuga.png',
    id: 'my-sprite',
    width: 160,
    height: 156,
    image_width: 1280,
    err: true
});
```
Some properties are optionall. **Below you can find a table with all available properties and methods. **

After object is created you can render it:

```javascript
example.render(1);
```
Or play animation loop:
```javascript
example.loop();
```

To see more examples please visit [demo page.](https://wisniewski94.github.io/sprites.js/)

## Browser Support
Sprite.js is based on HTML5 Canvas API and requestAnimationFrame which are partially/not supported in Opera Mini Browsers at the moment. These functionalities are the core of this library and cannot be replaced in order to work with older browsers or Opera Mini.

|IE 10+|Edge 12+|Firefox 11+|
|:---:|:---:|:---:|
|Chrome 22+|Safari 6+|Opera 15+|

## Options
These options can be used only on sprite initialization:

|function|property|default|optional|type|description|
|---|---|---|---|---|---|
|new Sprite|src|undefined|no|string|Path to image|
|new Sprite|id|undefined|no|string|Selects canvas tag which is going to be used for animation|
|new Sprite|width|undefined|no|number|Single frame width|
|new Sprite|height|undefined|no|number|Single frame height|
|new Sprite|image_width|src width|yes|number|You can use this option to crop or extend image on right side. Keep in mind that this property is used to calculate amount of frames.|
|new Sprite|err|false|yes|boolean|Errors will print in developers tools console|
|.play()|from|1|yes|number|First frame of the sequence.|
|.play()|to|last frame|yes|number|Last frame of sequence.|
|.play()|fps|60|yes|number|Frames per second.|
|.play()|n|1|yes|number|Amount of loops. By default sequence is played once. If set to 0 then loop is going to be infinite.
|.play()|step|none|yes|function|Returns a number of current frame in callback.|
|.play()|loop|none|yes|function|Returns a number of current loop (fires at last frame) in callback.|
## Methods
```javascript
// In order to skip parameter type null or undefined. Instead of specific parameter default value will be used.

/**
 * Play animation. You can set range of animation, fps, and number of passes (loops). Two callbacks available in object parameter.
 * @param {object} 
 */
this.play({
//Check settings table
});

/**
 * Return current frame.
 * @return {number} -> frameIndex.
 */
this.frame();

/**
 * Pause animation.
 */
this.pause();

/**
 * Reset animation and optionally jump to specific frame.
 * @param {number} to -> Default 1.
 */
this.reset(to);
```

## Contribute
This plugin is made by community for community. If you want to help develop this project you can do it by:

[Report a bug](https://github.com/wisniewski94/sprites.js/issues)

[Ask for a feature](https://github.com/wisniewski94/sprites.js/issues)

[Submit a pull request](https://github.com/wisniewski94/sprites.js/pulls)

### Donate
<a href='https://pledgie.com/campaigns/34691'><img alt='Click here to lend your support to: Open Source Projects - wisniewski94 and make a donation at pledgie.com !' src='https://pledgie.com/campaigns/34691.png?skin_name=chrome' border='0' ></a>

If you find this project useful you can also consider buying me a small coffee :)


HISTORY.MD
1.2.0
===============
* Changed `.play()` method parameter from arguments to object.
* Added `step` and `loop` callbacks to `.play()` method.
* Fixed bug where animation was stacking if `err` was set to `false`.
* Fixed bug related with `from` and `to` options in `.play()` method.
* Docs update and cleanup.

1.1.0
===============
* Code cleanup.
* Removed `.render()` method. `.reset()` does same job.
* Simplified errors and removed them from global scope.

1.0.1
===============
* Initial release.


sprite.js
'use strict';

function Sprite(opt) {
    //private settings
    var cvs     = document.getElementById(opt.id),
        x       = cvs.width = opt.width,
        y       = cvs.height = opt.height,
        ctx     = cvs.getContext('2d'),
        img     = new Image(),
        frames  = opt.image_width ? (opt.image_width / x) : (function (){throw missing}());
    
    //config
    var index   = 0,
        frame   = -1,
        time    = null,
        playing = false,
        err     = opt.err,
        playErr = new Error('Animation is already playing'),
        rangeErr= new RangeError('Parameter must be between 1 and ' + frames),
        missing = new Error('Missing property'),
        rAF, count, paused;
    
    img.src = opt.src;
    ctx.drawImage(img, index * x, 0, x, y, 0, 0, x, y);

    this.play = function (e) {
        if (playing && err) throw playErr;
        if (e.to > frames && err) rangeErr;
        if (e.from < 1 && err) throw rangeErr;
        if (e.fps == null) e.fps = 60;
        if (e.from == null) e.from = 1;
        if (e.to == null) e.to = frames;
        if (e.n == null) e.n = 1;
        if (paused == null || !paused || count !== count || count == null) {
            if (e.to > e.from) count = 0;
            if (e.from > e.to) count = 1;
        }
        var delay = 1000 / e.fps;
        index = e.from - 1;
        e.from--;
        e.to--;
        playing = true;
        rAF = requestAnimationFrame(play.bind(null, e.from, e.to, e.n, delay, e.step, e.loop));
    }
    
    this.pause = function () {
        if (!playing && err) throw playErr;
        cancelAnimationFrame(rAF);
        playing = false;
        paused = true;
        frame = -1;
    };
    
    this.reset = function (to) {
        if (to == null) to = 0;
        cancelAnimationFrame(rAF);
        playing = false;
        frame = -1;
        index = to - 1;
        ctx.clearRect(0, 0, x, y);
        ctx.drawImage(img, (--to) * x, 0, x, y, 0, 0, x, y);
    };
    
    this.frame = function() {
        return index + 1;
    };
    
    var play = function (from, to, n, delay, callback1, callback2, timestamp) {
        rAF = requestAnimationFrame(play.bind(null, from, to, n, delay, callback1, callback2));
        if (time === null) time = timestamp;
        var current = Math.floor((timestamp - time) / delay);
        if (current > frame) {
            frame = current;
            if (count >= n && index == to && n !== 0) {
                playing = false;
                paused = false;
            }
            ctx.clearRect(0, 0, x, y);
            ctx.drawImage(img, index * x, 0, x, y, 0, 0, x, y);
            if (to > from) {
                if (index >= to) {
                    typeof callback1 === 'function' && callback1(index + 1);
                    index = from;
                } else {
                    typeof callback1 === 'function' && callback1(index + 1);
                    index++;
                }
                if (index == to) {
                    typeof callback2 === 'function' && callback2(count);
                    count++;
                }
            } else if (from > to) {
                if (index <= to) {
                    typeof callback1 === 'function' && callback1(index + 1);                    
                    index = from;
                } else {
                    typeof callback1 === 'function' && callback1(index + 1);
                    index--;
                }
                if (index == from) {
                    typeof callback2 === 'function' && callback2(count);
                    count++;
                }
            }// else if

        } //current > frame
        
    };
}


sprites.js
'use strict';

function Sprite(opt) {
    //private settings
    var cvs     = document.getElementById(opt.id),
        x       = cvs.width = opt.width,
        y       = cvs.height = opt.height,
        ctx     = cvs.getContext('2d'),
        img     = new Image(),
        frames  = opt.image_width ? (opt.image_width / x) : (function (){throw missing}());
    
    //config
    var index   = 0,
        frame   = -1,
        time    = null,
        playing = false,
        err     = opt.err,
        playErr = new Error('Animation is already playing'),
        rangeErr= new RangeError('Parameter must be between 1 and ' + frames),
        missing = new Error('Missing property'),
        rAF, count, paused;
    
    img.src = opt.src;
    ctx.drawImage(img, index * x, 0, x, y, 0, 0, x, y);

    this.play = function (e) {
        if (playing && err) {console.error(playErr); return;};
        if (e.to > frames && err) {console.error(rangeErr); return;};
        if (e.from < 1 && err) {console.error(rangeErr); return;};
        if (e.fps == null) e.fps = 60;
        if (e.from == null) e.from = 1;
        if (e.to == null) e.to = frames;
        if (e.n == null) e.n = 1;
        if (paused == null || !paused || count !== count || count == null) {
            if (e.to > e.from) count = 0;
            if (e.from > e.to) count = 1;
        }
        var delay = 1000 / e.fps;
        index = e.from - 1;
        e.from--;
        e.to--;
        playing = true;
        rAF = requestAnimationFrame(play.bind(null, e.from, e.to, e.n, delay, e.step, e.loop));
    }
    
    this.pause = function () {
        if (!playing && err) throw playErr;
        cancelAnimationFrame(rAF);
        playing = false;
        paused = true;
        frame = -1;
    };
    
    this.reset = function (to) {
        if (to == null) to = 0;
        cancelAnimationFrame(rAF);
        playing = false;
        frame = -1;
        index = to - 1;
        ctx.clearRect(0, 0, x, y);
        ctx.drawImage(img, (--to) * x, 0, x, y, 0, 0, x, y);
    };
    
    this.frame = function() {
        return index + 1;
    };
    
    var play = function (from, to, n, delay, callback1, callback2, timestamp) {
        rAF = requestAnimationFrame(play.bind(null, from, to, n, delay, callback1, callback2));
        if (time === null) time = timestamp;
        var current = Math.floor((timestamp - time) / delay);
        if (current > frame) {
            frame = current;
            if (count >= n && index == to && n !== 0) {
                playing = false;
                paused = false;
            }
            ctx.clearRect(0, 0, x, y);
            ctx.drawImage(img, index * x, 0, x, y, 0, 0, x, y);
            if (to > from) {
                if (index >= to) {
                    typeof callback1 === 'function' && callback1(index + 1);
                    index = from;
                } else {
                    typeof callback1 === 'function' && callback1(index + 1);
                    index++;
                }
                if (index == to) {
                    typeof callback2 === 'function' && callback2(count);
                    count++;
                }
            } else if (from > to) {
                if (index <= to) {
                    typeof callback1 === 'function' && callback1(index + 1);                    
                    index = from;
                } else {
                    typeof callback1 === 'function' && callback1(index + 1);
                    index--;
                }
                if (index == from) {
                    typeof callback2 === 'function' && callback2(count);
                    count++;
                }
            }// else if

        } //current > frame
        
    };
}
